trace-utils/src/tracer_payload.rs

```diff
@@ -1,39 +1,132 @@
+        }
+    }
+
+    pub fn set_top_level_spans(&mut self) {
+        match self {
+            TraceCollection::TraceChunk(traces) => {
+                for chunk in traces.iter_mut() {
+                    trace_utils::compute_top_level_span(chunk);
+                }
+            }
+            TraceCollection::V07(_) => todo!("set_top_level_spans not implemented for v07"),
+        }
+    }
+
+    /// Remove spans and chunks only keeping the ones that may be sampled by the agent
+    pub fn drop_chunks(&mut self) -> (usize, usize) {
+        let mut dropped_p0_traces = 0;
+        let mut dropped_p0_spans = 0;
+
+        match self {
+            TraceCollection::TraceChunk(traces) => {
+                traces.retain_mut(|chunk| {
+                    // List of spans to keep even if the chunk is dropped
+                    let mut sampled_indexes = Vec::new();
+                    for (index, span) in chunk.iter().enumerate() {
+                        // ErrorSampler
+                        if span.error == 1 {
+                            // We send chunks containing an error
+                            return true;
+                        }
+                        // PrioritySampler and NoPrioritySampler
+                        let priority = span.metrics.get(SAMPLING_PRIORITY_KEY);
+                        if trace_utils::has_top_level(span)
+                            && (priority.is_none() || priority.is_some_and(|p| *p > 0.0))
+                        {
+                            // We send chunks with positive priority or no priority
+                            return true;
+                        }
+                        // SingleSpanSampler and AnalyzedSpansSampler
+                        else if span
+                            .metrics
+                            .get(SAMPLING_SINGLE_SPAN_MECHANISM)
+                            .is_some_and(|m| *m == 8.0)
+                            || span.metrics.contains_key(SAMPLING_ANALYTICS_RATE_KEY)
+                        {
+                            // We send spans sampled by single-span sampling or analyzed spans
+                            sampled_indexes.push(index);
+                        }
+                    }
+                    dropped_p0_spans += chunk.len() - sampled_indexes.len();
+                    if sampled_indexes.is_empty() {
+                        // If no spans were sampled we can drop the whole chunk
+                        dropped_p0_traces += 1;
+                        return false;
+                    }
+                    let sampled_spans = sampled_indexes
+                        .iter()
+                        .map(|i| std::mem::take(&mut chunk[*i]))
+                        .collect();
+                    *chunk = sampled_spans;
+                    true
+                });
+            }
+            TraceCollection::V07(_) => todo!("drop_chunks not implemented for v07"),
```

Comment:
We haven't formalized a no-panic policy yet, but I think we should at least refrain from adding new panics. If it's easier to address this in a follow-up PR (I assume we'll have to add some new error types and change code to handle those errors, so not a trivial amount of work)that is ok. If that's the case can you add a `TODO` comment with a jira ID anywhere you've added new panics?